torchtune/utils/_distributed.py

```diff
@@ -297,3 +299,69 @@ def lora_wrap_fsdp(module: nn.Module, recurse: bool, **kwargs):
         return isinstance(module, tuple(modules_to_wrap))
 
     return lora_wrap_fsdp
+
+
+def load_from_full_state_dict(
+    model: torch.distributed._composable.fsdp.FSDP,
+    full_sd: Dict[str, Any],
+    device: torch.device,
+    is_rank_zero: bool,
+):
+    meta_sharded_sd = model.state_dict()
+    sharded_sd = {}
+    for param_name, full_tensor in full_sd.items():
+        sharded_meta_param = meta_sharded_sd.get(param_name)
+        if is_rank_zero:
+            full_tensor = full_tensor.detach().to(device)
+        else:
+            full_tensor = torch.empty(
+                sharded_meta_param.size(),
+                device=device,
+                dtype=sharded_meta_param.dtype,
+            )
+        torch.distributed.broadcast(full_tensor, src=0)
+        sharded_tensor = distribute_tensor(
+            full_tensor, sharded_meta_param.device_mesh, sharded_meta_param.placements
+        )
+        sharded_sd[param_name] = nn.Parameter(sharded_tensor)
+    model.load_state_dict(sharded_sd, strict=False, assign=True)
+
+
+def get_full_model_state_dict(
+    model: torch.distributed._composable.fsdp.FSDP,
+    is_rank_zero: bool,
+) -> Dict[str, Any]:
+    sharded_sd = model.state_dict()
+    cpu_state_dict = {}
+    for param_name, sharded_param in sharded_sd.items():
+        full_param = sharded_param.full_tensor()
+        if is_rank_zero:
+            cpu_state_dict[param_name] = full_param.cpu()
+        else:
+            del full_param
+    return cpu_state_dict
+
+
```

Comment:
@awgu I made 2 assumptions about `opt.state_dict()` when converting DTensor into full tensor. Let me know if it is too aggresive
* `sharded_sd["param_groups"]` are the same across ranks. no need to communicate
* `sharded_sd["state"]`: for AdamW, `key=step` and `value` is plain Tensor. `key=exp_avg/exp_avg_sq`, value is `DTensor`. We only need to communicate on DTensor. It might not work for Apex optimizers

Reply:
This seems reasonable to me since these assumptions follow the SPMD programming model that `DTensor`/FSDP generally assume. If the `"param_groups"` were different across ranks, I am not sure how to coalesce them.

cc: @fegin @wz337  who have more experience with this

Separately, I think we should discuss with the torchtune folks why they prefer full optimizer state dict. For model, I can buy that we want to export the model state dict for downstream use, but I would be curious if there are cases where we want to export the optimizer state dict and resume training e.g. from a single-GPU setup.

Reply:
@weifengpy Why not just use `torch.distributed.checkpoint.state_dict.get_state_dict`? That will save all the implementations here.